libgadget/forcetree.h

```diff
@@ -20,10 +22,19 @@ struct NODE
         unsigned int DependsOnLocalMass :1;  /* Intersects with local mass */
         unsigned int MixedSofteningsInNode:1;  /* Softening is mixed, need to open the node */
         unsigned int NodeIsDirty :1; /*Node is a toplevel node containing local mass, and its moments need updating*/
+        unsigned int InternalNode :1; /* If true, node contains only other nodes.
+                                       * Otherwise, node contains particles and pseudo particles*/
     } f;
     union
     {
-        int suns[8];		/*!< temporary pointers to daughter nodes */
+        struct
+        {
+            /*!< temporary pointers to daughter nodes or daughter particles. */
+            int suns[NMAXCHILD];
+            /* Number of daughter particles if node contains particles.
+             * During treebuild >= (1<<16) if node contains nodes.*/
+            int noccupied;
```

Comment:
What about splitting it to two short ints, one for noccupried, and one for InternalNode?

Reply:
Or hide 1<<16 to a macro ?

Reply:
I want to keep InternalNode around so I can use it in the future in the treewalk code to reduce branches (not sure how yet). Currently 1<<16 is only used during the first phase of the treebuild to make it one less atomic read. It's a little ugly but it really means "a number much larger than NCHILDMAX" and I think that's ok. All the macros I can think of are excessively magic because of the atomic-ness.